### PR TITLE
api: Use requests.Session

### PR DESCRIPTION
**Pull request by dehnert**
Using requests.Session allows the requests library to reuse HTTP connections,
which is potentially helpful for performance. This fixes python-zulip-api #3.

Some notes:
* I'm currently lazily initializing the Session object. This has the advantage if somebody constructs a client and then modifies some field before making any requests, it should work as they probably expect. It has the disadvantage of probably not being threadsafe (although the client as a whole might not be either), not to mention whatever style views people have.
* It appears to pass lint. AFAIK, that's it for "testing" the API -- I think the bot tests mock it out? (But those tests do pass.)
* I ran `mypy` (though not the version that Zulip has pinned). Beyond some warnings about stubs that I think are preexisting, I get some errors about request.Session's field typing. I think typeshed is just wrong; https://github.com/python/typeshed/issues/1093 certainly does not leave me with much faith in the stubs. Comparing [Session's stubs](https://github.com/python/typeshed/blob/master/third_party/2and3/requests/sessions.pyi) to [.request's stubs](https://github.com/python/typeshed/blob/master/third_party/2and3/requests/api.pyi) the former is certainly more restrictive, and I don't believe the stubs' typing for [verify](http://docs.python-requests.org/en/master/user/advanced/#ssl-cert-verification) (which really does seem to take a string) or auth (which does seem to take [an object](http://docs.python-requests.org/en/master/user/advanced/#custom-authentication) -- though in fairness it looks like passing a (user,pass) tuple should work too, and the stubs support it). I'm not sure what to do about this, especially given that the API seems annotated but not currently being tested.

I think this is sane to merge, although you might reasonably want me to change how I handled these notes (or have other objections).

### CODE REVIEW

File: zulip/zulip/__init__.py [hunk 1]
```diff
@@ -327,6 +351,8 @@ def do_api_query(self, orig_request, url, method="POST", longpolling=False, file
         for f in files:
             req_files.append((f.name, f))
 
+        self._init_session()
```
Comment (timabbott) [inline zulip/zulip/__init__.py]: maybe `self.ensure_session()` would be a better name here?

**Comment by timabbott** [comment]
The code looks good; posted one quick comment.  Some thoughts on your notes:
* I think threadsafety is something I'd rather handle via locking/whatever if we find a use for it.  
* For typeshed, I'd open an issue in Typeshed and use `# type: ignore # link to issue` to silence the warnings in our code and reference it so it's clear why and when we can clean it up.  We'll be type-checking the API again eventually.
* Correct, this has no real test suite in this project yet (there's a bit more of one in the server project in the form of `tools/test-api` there).  Might be worth running that against this project, because I think there's a good chance that its mocking strategy won't work with this.

**Comment by dehnert** [comment]
Thanks for the feedback. I'm traveling this weekend, but I'll try to deal early next week. 

On July 20, 2017 4:07:54 PM PDT, Tim Abbott <notifications@github.com> wrote:
>The code looks good; posted one quick comment.  Some thoughts on your
>notes:
>* I think threadsafety is something I'd rather handle via
>locking/whatever if we find a use for it.  
>* For typeshed, I'd open an issue in Typeshed and use `# type: ignore #
>link to issue` to silence the warnings in our code and reference it so
>it's clear why and when we can clean it up.  We'll be type-checking the
>API again eventually.
>* Correct, this has no real test suite in this project yet (there's a
>bit more of one in the server project in the form of `tools/test-api`
>there).  Might be worth running that against this project, because I
>think there's a good chance that its mocking strategy won't work with
>this.  
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/zulip/python-zulip-api/pull/18#issuecomment-316855358

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by timabbott** [comment]
Cool

**Comment by timabbott** [comment]
@dehnert quick ping on this :)

**Comment by dehnert** [comment]
I fixed these comments last night, except that I haven't been able to run `test-api` yet -- my Zulip proper environment (without these changes) is failing to run the tests. On the typeshed front, I submitted https://github.com/python/typeshed/pull/1504 upstream and referenced it from the code.

**Comment by timabbott** [comment]
@eeshangarg do you have time to quickly run `test-api` against an updated version `python-zulip-api` that contains this code?

**Comment by rht** [comment]
[Passed in Travis](https://travis-ci.org/rht/zulip/builds/262537254) ! (I
suppose a fresh dev build doesn't have the hysteresis of accumulated errs from
past experimentation.)

Tried porting a subset of the backend test in zulip/zulip
https://github.com/rht/python-zulip-api/commit/2f967458d6171f1abb0ceb205669937aa558fbeb
just enough to run `./tools/test-api` but to no avail since the
`./tools/provision` from the zulip/zulip repo would still fetch the old version
instead of the hash being tested. So I treaded the path of least loc-patch by
patching the requirements in zulip/zulip.

**Comment by timabbott** [comment]
@rht for the zulip/zulip issue, you just need to update `requirements.txt` to point to the Git URL of this branch in the relevant fork, and then you can provision against this version.

**Comment by timabbott** [comment]
Merged, thanks @dehnert!  I think it'll be easier to test by hand once it's merged and we can just include it in the server repo.  It seems pretty unlikely that it'll fail in a nontrivial way.

**Comment by dehnert** [comment]
Thanks folks!

**Pull request closed without merging**